### PR TITLE
RavenDB-20877: Part 2 - Refactoring

### PR DESCRIPTION
**Pull request by redknightlois**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-22261

### Additional description

Part 2 that includes many refactoring and code cleanup. 

### Type of change

- [ ] Bug fix
- [ ] Regression bug fix
- [x] Optimization
- [ ] New feature

### How risky is the change?

- [ ] Low 
- [x] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [ ] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [x] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor

- [ ] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

### CODE REVIEW

**Comment by redknightlois** [comment]
This PR includes https://github.com/ravendb/ravendb/pull/19038 and show be reviewed after merging.

**Comment by redknightlois** [comment]
This one is ready to be reviewed.

File: src/Voron/Data/Fixed/FixedSizeTree.cs [hunk 1]
```diff
@@ -108,52 +110,60 @@ private void ThrowScopeAlreadyOpen()
...
         }
 
-        public void RepurposeInstance(Slice treeName, bool clone)
+        public static bool TryRepurposeInstance(FixedSizeTree<TVal> tree, Slice treeName, bool clone)
         {
-            new DirectAddScope(this).Dispose();// verifying that we aren't holding a ptr out
+            new DirectAddScope(tree).Dispose();// verifying that we aren't holding a ptr out
 
+            // Setting the name of the tree has to happen before returning even if the return is null
+            // TODO: Make sure that is no longer the case by prohibiting the direct creation of trees.
```
Comment (ayende) [inline src/Voron/Data/Fixed/FixedSizeTree.cs:118]: What remains to be done here?

File: src/Voron/Data/Tables/Table.cs [hunk 2]
```diff
@@ -2442,7 +2443,8 @@ public TableReport GetReport(bool includeDetails, StorageReportGenerator generat
 
             report.AddData(ActiveDataSmallSection, includeDetails);
 
-            report.AddPreAllocatedBuffers(TablePageAllocator, includeDetails);
+            var allocator = new NewPageAllocator(_tx.LowLevelTransaction, _tableTree);
```
Comment (ayende) [inline src/Voron/Data/Tables/Table.cs]: Why are you creating a new one here? We now no longer create it by default, but there is an `EnsureTablePageAllocator` that you can call instear.

File: src/Voron/Data/Tables/TableValueReader.cs [hunk 3]
```diff
@@ -1,48 +1,112 @@
...
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
 using System.Text;
+using Sparrow;
 using Sparrow.Compression;
 
 namespace Voron.Data.Tables
 {
-    public readonly unsafe struct TableValueReader
+    public unsafe interface ITableValueReader
```
Comment (ayende) [inline src/Voron/Data/Tables/TableValueReader.cs]: Why do we need an interface here?

Comment (ayende) [reply src/Voron/Data/Tables/TableValueReader.cs]: For the `N1`, `N2`, etc?

**Review by ayende — APPROVED** [review]

File: src/Voron/Data/Tables/Table.cs (hunk 2 above)
Comment (redknightlois) [reply src/Voron/Data/Tables/Table.cs]: Good catch, no longer needed. It was a left over.

File: src/Voron/Data/Tables/TableValueReader.cs (hunk 3 above)
Comment (redknightlois) [reply src/Voron/Data/Tables/TableValueReader.cs]: Yes, that's how we achieve code specialization.

File: src/Voron/Data/Fixed/FixedSizeTree.cs (hunk 1 above)
Comment (redknightlois) [reply src/Voron/Data/Fixed/FixedSizeTree.cs:118]: At this point in time, that is still missing. No longer there in Part3.

**Comment by redknightlois** [comment]
The test `RavenDB_22811.CoraxIndexMustNotThrowErrorsOnNonIndexedComplexFieldAfterSideBySideIndexing` which is in error is in failure before this PR. 

EDIT: Fixed by https://github.com/ravendb/ravendb/pull/19169

**Comment by ppekrol** [comment]
@redknightlois can you check the failing test? Looks very related

**Comment by redknightlois** [comment]
Can you rerun on Debug? Can't repro locally.
This is release version rerun: https://jenkins.hibernatingrhinos.com/job/PR_Tests/12106/

**Comment by redknightlois** [comment]
Apparently error is unrelated as it triggers without the PR too. https://issues.hibernatingrhinos.com/issue/RavenDB-22859/Jalchr3.Streamingdocumentswillrespectthesortingorder

**Comment by ppekrol** [comment]
And this one https://jenkins.hibernatingrhinos.com/job/PR_Tests/12083/testReport/junit/SlowTests.Client.Counters/BulkInsertCounters/IncrementManyCounters_options___DatabaseMode___Single___SearchEngineMode___Lucene__counterCount__100000_/ ?

**Comment by redknightlois** [comment]
That last is not reproducible on my end. Which is weird. I will try a few things, but seems like it is the same kind of timing issue as the other one. 

`Jalchr3` was pretty difficult to reproduce locally.
image
That's roughly an hour of execution.

File: src/Sparrow/Json/BlittableJsonReaderObject.cs [hunk 4]
```diff
@@ -1471,7 +1462,7 @@ public static void AssertNoModifications(BlittableJsonReaderObject data, string
 
         public bool Contains(LazyStringValue propertyName)
         {
-            AssertContextNotDisposed();
+            ThrowIfDisposedOnDebug(this);
```
Comment (ppekrol) [inline src/Sparrow/Json/BlittableJsonReaderObject.cs:1465]: it is super hard to review if you are using static usings

File: src/Voron/VoronConfiguration.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+        // https://alexandrnikitin.github.io/blog/jit-optimization-static-readonly-to-const/
+        
+        public static readonly bool FailFastForStability;
+
+        static VoronConfiguration()
+        {
+#if DEBUG
+            FailFastForStability = true;
+#else
+            if (Environment.GetEnvironmentVariable("RAVENDB_Voron_FailFast")?.ToLowerInvariant() == "true")
```
Comment (ppekrol) [inline src/Voron/VoronConfiguration.cs:18]: Our env looks like RAVEN_HTTP_POOLEDCONNECTIONLIFETIMEINSEC, can you adjust? Can you fix as well the ones in AdvInstructionSet?

Comment (ppekrol) [reply src/Voron/VoronConfiguration.cs:18]: bool.TryParse would be better? You do not have to deal with whitespaces then?

**Review by ppekrol — APPROVED** [review]

**Comment by ppekrol** [comment]
Merging this, please open another PR with adjustments

File: src/Sparrow/Json/BlittableJsonReaderObject.cs (hunk 4 above)
Comment (redknightlois) [reply src/Sparrow/Json/BlittableJsonReaderObject.cs:1465]: The code is much better with static usings for throwing instructions and runtime asserts. Same happens when using intrinsics, but I do agree it is the only place where it actually makes sense.

**Pull request merged**